content/macro-annotations.md

```diff
@@ -0,0 +1,349 @@
+As we do not want to allow a macro annotation to see new definitions added by another macro annotation, we will enter the symbol at the end of the inlining phase.
+
+
+##### Expansion order
+
+A definition might be annotated with multiple macro annotations, for example: `@log @memoize def fib(n: Int) = ...`.
+We need to expand them in some reliable order, either left to right or right to left.
+These two orders are equivalently acceptable, as a tiebreaker we use the right to left that Scala 2 macro annotations chose.
+It is not encouraged to write one macro annotation relying on another that annotates the same definition.
+If so, the user could write a new macro annotation that does this transformation calling into the implementation of the other macro annotations.
+If we have macro annotated parameters we start from the last one, to follow the same order as the annotations of the definition itself.
+
+We can also have a macro annotation on a definition that is nested in a macro annotated definition.
+We can also follow the Scala 2 design and expand the outer macro annotation first, then the inner ones.
+
+#### Restrictions
+
+The `MacroAnnotation.transform` methods imposed some restrictions on what can be returned in the list of definitions.
+
+* The list must contain the transformed definition (a definition with the same symbol as the original definition).
+* If a companion is modified, it must be returned in the list of definitions and must have the original symbol.
+* All definitions in the result must have the same owner. The owner can be recovered from `Symbol.spliceOwner`.
+* Special case: an annotated top-level `def`, `val`, `var`, `lazy val` can return a `class`/`object`
+                definition that is owned by the package or package object.
+* Can not return a `type`. At this abstraction level, they are not really useful.
+* Annotated top-level `class`/`object` can not return top-level `def`, `val`, `var`, `lazy val`.
+* Trees must be well-formed TASTy definition. `-Xcheck-macros` and `-Ycheck` will help to enforce this.
+
+
+
+### Compatibility
+The introduction of the `MacroAnnotation` trait does not affect binary nor TASTy compatibility.
+
+Crucially, macro annotations cannot break TASTy compatibility because they are evaluated after pickling.
+They should also not break binary compatibility as newly added definitions by macro annotations are assumed to only be used from within the macro expanded code.
+
+Macro annotations can be used to patch binary incompatibilities by generating an old definition that is not generated anymore.
+
+
+### Feature Interactions
+
+
+#### Typer
+As we transform the definitions after pickling, the typer only needs to be able to type the reference to the macro annotation.
+
+#### Separate/incremental compilation
+As we transform the definitions after pickling, separate/incremental compilation is not affected by the macro annotations.
```

Comment:
What if a macro annotated member of a `trait` generates a new `private val/var/lazy val` in the trait? That would still affect classes extending that trait, in a way that incremental compilation should be aware of.

Reply:
I am not sure how we would support these. They might need to be disallowed.